app/code/core/Mage/Adminhtml/Block/Widget/Grid.php

```diff
@@ -970,29 +970,29 @@ public function _exportIterateCollection($callback, array $args)
         $originalCollection = $this->getCollection();
         $count = null;
         $page  = 1;
-        $lPage = null;
-        $break = false;
 
-        while ($break !== true) {
+        while (true) {
             $collection = clone $originalCollection;
             $collection->setPageSize($this->_exportPageSize);
             $collection->setCurPage($page);
             $collection->load();
-            if (is_null($count)) {
-                $count = $collection->getSize();
-                $lPage = $collection->getLastPageNumber();
-            }
-            if ($lPage == $page) {
-                $break = true;
+
+            $count = $collection->count();
+
+            if ($count < $this->_exportPageSize) {
```

Comment:
I didn't test this. But assuming `$originalCollection` has 1001 items, given that `protected $_exportPageSize = 1000;`, it will require 2 iterations. On the second iteration, `$count` will be 1, but it won't get exported since it'll `break` here. So, we need `$break = true;` instead of `break;`. Or am I missing something.

Reply:
If I am right, `$count` can be init to `$this->_exportPageSize`:

```php
$originalCollection = $this->getCollection();
$count = $this->_exportPageSize;
//...
while ($count != $this->_exportPageSize) {
    //...
    $count = $collection->getSize(); // I prefer the standard method.
    // we can skip `if ($count < $this->_exportPageSize) {` and continue to export
}
```

Reply:
Your thoughts seems right. My mistake. I swear I tested it, we are using the patched version in our Magento instance. I will test it again.

Reply:
Ok I figured it out. There is two methods `getCsv` and `getCsvFile`. Only `getCsvFile` is using `_exportIterateCollection`. But in most grids, the `getCsv` is used, so when I was testing my code, I actually run the other method. The right place to test it is tax rates grid:

https://yourwebsite/admin/tax_rate/

Reply:
It is fixed now and tested.